Add `src/node/` and `src/wallet/` code to `node::` and `wallet::` namespaces

ryanofsky opened the pull request:
There are no code changes, this is just adding `namespace` and `using` declarations and `node::` or `wallet::` qualifiers in some places.

Motivations for this change are:

- To make it easier to see when node and wallet code is being accessed places where it shouldn't be. For example if GUI code is accessing node and wallet internals or if wallet and node code are referencing each other.
- To make source code organization clearer ([#15732](https://github.com/bitcoin/bitcoin/issues/15732)), being able to know that `wallet::` code is in `src/wallet/`, `node::` code is in `src/node/`, `init::` code is in `src/init/`, `util::` code is in `src/util/`, etc.

Reviewing with `git log -p -n1 -U0 --word-diff-regex=.` can be helpful to verify this is only updating declarations, not changing code.

ryanofsky committed:
Add src/node/* code to node:: namespace

ryanofsky committed:
Add src/wallet/* code to wallet:: namespace

ryanofsky commented on src/qt/psbtoperationsdialog.cpp:
```diff
@@ -17,6 +17,10 @@
 #include <iostream>
 
+using node::AnalyzePSBT;
+using node::BroadcastTransaction;
```
In commit "Add src/node/* code to node:: namespace" (733a629696ef3350e5c9c861ae94254a06cc920b)

Note: This line can be removed after #23499, since calling `node::BroadcastTransaction` from the bitcoin-gui process doesn't work in the multiprocess build (there is a segfault because the NodeContext pointer this tries to pass is null). This is a preexisting bug, independent of this PR.

ryanofsky commented:
Updated 768b695296e343da0163e07e25f08dfbc6f58e1d -> 8c85a48eac8a0a9512e754f29c173c2ce9e55308 ([`pr/names.1`](https://github.com/ryanofsky/bitcoin/commits/pr/names.1) -> [`pr/names.2`](https://github.com/ryanofsky/bitcoin/commits/pr/names.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/names.1..pr/names.2)) with tweaks to clean up and fix errors in dummy wallet and multiprocess builds https://cirrus-ci.com/task/5598684416049152 and https://cirrus-ci.com/task/6161634369470464

katesalazar commented on src/interfaces/init.h:
```diff
@@ -7,7 +7,9 @@
 
+namespace node {
 struct NodeContext;
+}  // namespace node
```
Some lines in pull 23492 remove one of the two spaces elsewhere.

lsilva01 reviewed (approved):
tACK 8c85a48
The use of namespaces makes the code more organized and clearer.

stratospher reviewed (commented):
ACK 8c85a48. The changes applied make the code more understandable.

ryanofsky commented:
Rebased 8c85a48eac8a0a9512e754f29c173c2ce9e55308 -> 3a2c4e28775d49da233d04bed03e5cb8fe9d7239 ([`pr/names.2`](https://github.com/ryanofsky/bitcoin/commits/pr/names.2) -> [`pr/names.3`](https://github.com/ryanofsky/bitcoin/commits/pr/names.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/names.2-rebase..pr/names.3)) due to conflict with #23524

ryanofsky reviewed (commented):
Rebased 3a2c4e28775d49da233d04bed03e5cb8fe9d7239 -> 6ea409f96ccc37c74998abad4cf56c4999b3ba8a ([`pr/names.3`](https://github.com/ryanofsky/bitcoin/commits/pr/names.3) -> [`pr/names.4`](https://github.com/ryanofsky/bitcoin/commits/pr/names.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/names.3-rebase..pr/names.4)) to remove uneeded using declaration after #23499 and switch to one space instead of two spaces before trailing namespace comments.
Rebased 6ea409f96ccc37c74998abad4cf56c4999b3ba8a -> d72688f16dd1056e3253663769034b100341664c ([`pr/names.4`](https://github.com/ryanofsky/bitcoin/commits/pr/names.4) -> [`pr/names.5`](https://github.com/ryanofsky/bitcoin/commits/pr/names.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/names.4-rebase..pr/names.5)) due to conflicts with #23591 and #21206

meshcollider commented:
Concept ACK. I realise this PR is going to conflict with pretty much every move from rpcwallet. I'll prioritise reviewing this.

ryanofsky commented:
Rebased d72688f16dd1056e3253663769034b100341664c -> 0818cef3c5f6c310a953d3d90b81181a65a08dde ([`pr/names.5`](https://github.com/ryanofsky/bitcoin/commits/pr/names.5) -> [`pr/names.6`](https://github.com/ryanofsky/bitcoin/commits/pr/names.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/names.5-rebase..pr/names.6)) due to conflicts with #23602, #23640, #23155

ryanofsky commented:
> Concept ACK. I realise this PR is going to conflict with pretty much every move from rpcwallet. Please let me know if you'd prefer to have this merged first or after the moves, I'll prioritise reviewing this.

Thanks, the only changes in this PR are adding `namespace` and `using` lines in many places and `node::` `wallet::` prefixes in a few places, so it's not hard to review or maintain this, and it doesn't matter very much if it is ordered before or after the rpc moves. If there's only 1 RPC move PR, then it especially doesn't matter. But if there are 10 RPC move PRs and gaps between them, then this will need to be rebased 10 times, and that could be a annoying if it requires re-acks.

ryanofsky commented:
Rebased 0818cef3c5f6c310a953d3d90b81181a65a08dde -> 64c8123c6e36f42767383bd5725299a7a2b9e964 ([`pr/names.6`](https://github.com/ryanofsky/bitcoin/commits/pr/names.6) -> [`pr/names.7`](https://github.com/ryanofsky/bitcoin/commits/pr/names.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/names.6-rebase..pr/names.7)) due to conflicts with #23647

ryanofsky commented:
Rebased 64c8123c6e36f42767383bd5725299a7a2b9e964 -> 65c6a10a11db40b4bcf4176752334cf2c1a274ca ([`pr/names.7`](https://github.com/ryanofsky/bitcoin/commits/pr/names.7) -> [`pr/names.8`](https://github.com/ryanofsky/bitcoin/commits/pr/names.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/names.7-rebase..pr/names.8)) due to conflict with #23667

meshcollider commented on src/qt/walletmodel.h:
```diff
@@ -32,7 +32,6 @@ class SendCoinsRecipient;
-class CCoinControl;
 class CKeyID;
 class COutPoint;
 class COutput;
```
This should be in the wallet namespace too I think

meshcollider reviewed (commented):
Concept ACK

Have only reviewed the wallet commit so far. I think you missed a `CWallet` in `dummywallet.cpp`.

meshcollider commented on src/node/blockstorage.cpp:
```diff
@@ -94,12 +95,14 @@ void CleanupBlockRevFiles()
 }
+} // namespace node
 
 std::string CBlockFileInfo::ToString() const
```
Would it be easier / cleaner to move this so it doesn't force a weird break in the namespace?

meshcollider reviewed (commented):
utACK 65c6a10a11db40b4bcf4176752334cf2c1a274ca modulo above comments

ryanofsky committed:
Move CBlockFileInfo::ToString method where class is declared

CBlockFileInfo class is declared in src/chain.h, so move ToString
definition to src/chain.cpp instead of src/node/blockstorage.cpp

ryanofsky reviewed (commented):
Rebased 65c6a10a11db40b4bcf4176752334cf2c1a274ca -> de0100b765a7bed45fa4e5d4fc24cc30cece957a ([`pr/names.8`](https://github.com/ryanofsky/bitcoin/commits/pr/names.8) -> [`pr/names.9`](https://github.com/ryanofsky/bitcoin/commits/pr/names.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/names.8-rebase..pr/names.9)) due to conflict with #20295, also making suggested changes

---

re: https://github.com/bitcoin/bitcoin/pull/23497#pullrequestreview-826179864

> I think you missed a `CWallet` in `dummywallet.cpp`.

Thanks, removed because the only thing that was using it (a `MakeWallet` stub) was unused

meshcollider commented:
utACK de0100b765a7bed45fa4e5d4fc24cc30cece957a

ryanofsky commented:
Rebased de0100b765a7bed45fa4e5d4fc24cc30cece957a -> 175127797f6f6c052e1bd210e174a1aa48a5e1c6 ([`pr/names.9`](https://github.com/ryanofsky/bitcoin/commits/pr/names.9) -> [`pr/names.10`](https://github.com/ryanofsky/bitcoin/commits/pr/names.10), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/names.9-rebase..pr/names.10)) due to conflict with #22019

ryanofsky commented:
@dongcarl do you think could review this? The PR just adds `namespace node { ... }` around src/node/ code and `namespace wallet { ... }` around src/wallet/ code and I think should be a quick review

MarcoFalke commented on src/index/base.cpp:
```diff
@@ -14,6 +14,8 @@
 #include <validation.h> // For g_chainman
 #include <warnings.h>
 
+using node::ReadBlockFromDisk;
```
What is the rule here? Randomly pick `using` or specify the namespace before the symbol?

ryanofsky replied:
> What is the rule here? Randomly pick `using` or specify the namespace before the symbol?

I don't know if there's a rule, but I think `using some_namespace::SomeSymbol;` statements are generally a bad thing, and should be used sparingly. They are bad because they can make it unclear where symbols are coming from, and can lead to ugly code when longer names and shorter names are used inconsistently.

In this PR, I'm trying to limit scope of `using` statements and only adding them locally in .cpp files, never in header files. I think most of the places where they are added are places where code organization should be improved, and that these will be removed when it is improved. For example, I'm working on porting src/index/ code to use `interfaces::Chain` and stop using node code and `node::` symbols. Carl and James and others are moving more validation code to `src/node/` which will make `using` statements in validation code go away. The `using` statements in RPC, bench, and test code could potentially be removed by having a `src/node/test/`directory analogous to the existing `src/wallet/test/` directory and a `src/node/rpc/` directory analagous to the existing `src/wallet/rpc/` directory. Or they could just be removed by adding `node::` prefixes to relevant types and symbols. I think either of these changes would be a win for clarity and organization, and the goal of this PR is to enable that.

ryanofsky replied:
> Carl and James and others are moving more validation code to `src/node/` which will make `using` statements in validation code go away.

Actually, I'm assuming too much here. Validation code doesn't have to move in order to avoid `using node::` statements  or `node::` prefixes. It could alternately become more modular and stop depending directly on `node::` types and symbols. It could move someplace other than src/node/, like src/kernel/, or not move at all. The number of using statements and node:: references goes down in any case.

dongcarl replied:
Agree that the `using` statements should be used sparingly.

If I understand correctly, the hope is that the `using` statements that you have right now will be removed in the future as we find better interfaces/boundaries between our modules? I can see how that would work for src/index/ code and `interfaces::Chain`, but I'm wondering how we'd get rid of cases like the various places with `using node::NodeContext;`, or perhaps these cases are ones where we want to keep the `using` statements?

ryanofsky replied:
> If I understand correctly, the hope is that the `using` statements that you have right now will be removed in the future as we find better interfaces/boundaries between our modules?

Yes, but I think this change is good whether or not that happens. The `using` statements just show places where module boundaries are being crossed. In general it's better if modules are self-contained and boundaries are crossed in fewer places. But it's useful in any case to be able to see the places where boundaries are crossed and not have hidden dependencies.

> I can see how that would work for src/index/ code and `interfaces::Chain`, but I'm wondering how we'd get rid of cases like the various places with `using node::NodeContext;`, or perhaps these cases are ones where we want to keep the `using` statements?

Just if you move the code using `node::NodeContext` *inside* the `node` namespace, then you no longer need `node::` prefixes or `using` statements to reference it. This is what I was referring to with the `src/node/rpc/` `src/node/test/` idea above, but that's just one possible way the node module could be more self-contained.

(Also `NodeContext` is kind of a weird case because it's the one node identifier that begins with word "node". I was thinking of making a followup PR that would remove all `using node::NodeContext` and rename `node::NodeContext` to `node::Context` and rename `wallet::WalletContext` to `wallet::Context`. Would be purely an aesthetic change, though, and low on the priority list.)

dongcarl commented:
Concept ACK, I like that this forces us to be somewhat more explicit.

Would like to wrap my head around `using` vs fully specifying in https://github.com/bitcoin/bitcoin/pull/23497#discussion_r772423809 though, let's discuss in that thread.

ryanofsky commented:
Rebased 91aaee1be3d7a63c764db61570f06f4a3daa87e2 -> 4decc9a53dcf36ec93b75cb2f029d07e008e00d0 ([`pr/names.11`](https://github.com/ryanofsky/bitcoin/commits/pr/names.11) -> [`pr/names.12`](https://github.com/ryanofsky/bitcoin/commits/pr/names.12), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/names.11-rebase..pr/names.12)) due to conflicts with #23842 and #23736

meshcollider commented:
Would this resolve the TODO on [line 28 of blockstorage.cpp](https://github.com/bitcoin/bitcoin/blob/327e2691f6e0498aa868af965bb4a951c5be17c5/src/node/blockstorage.cpp#L28)?

ryanofsky commented:
> Would this resolve the TODO on [line 28 of blockstorage.cpp](https://github.com/bitcoin/bitcoin/blob/327e2691f6e0498aa868af965bb4a951c5be17c5/src/node/blockstorage.cpp#L28)?
> 
> EDIT: actually, don't think so

It's arguable, but I figured I'd be proactive and remove the TODO in a new commit. I can drop the TODO or try to clarify it if someone thinks it should be kept.

Added 1 commits 4decc9a53dcf36ec93b75cb2f029d07e008e00d0 -> 76e3dda368811f3a49e0b6fd52cff612cbb3fd00 ([`pr/names.12`](https://github.com/ryanofsky/bitcoin/commits/pr/names.12) -> [`pr/names.13`](https://github.com/ryanofsky/bitcoin/commits/pr/names.13), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/names.12...pr/names.13))

MarcoFalke commented:
It is ok to remove the TODO. It still needs to be fixed (symbols should be private), but using a namespace is only one way to fix it. Alternatives are: Using `static` or `private:`.

MarcoFalke commented:
Done in #23974 (draft)

ryanofsky commented:
Rebased 76e3dda368811f3a49e0b6fd52cff612cbb3fd00 -> 4ee68619fad12f99a8c8b41a9f192503ad7789ea ([`pr/names.13`](https://github.com/ryanofsky/bitcoin/commits/pr/names.13) -> [`pr/names.14`](https://github.com/ryanofsky/bitcoin/commits/pr/names.14), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/names.13-rebase..pr/names.14)) due to conflicts with #23936 and #23581

MarcoFalke commented:
If you remove the last commit, a conflict with #23974 can be avoided

achow101 commented:
ACK e5b6aef61221b621ad77b5f075a16897e08835bf

MarcoFalke commented:
Concept ACK e5b6aef61221b621ad77b5f075a16897e08835bf 🍨

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Concept ACK e5b6aef61221b621ad77b5f075a16897e08835bf 🍨
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjJRwv/Ra0FnQ2CgmimyjQbDwPhXhsHl4FbFarePo1urtKhZ/yzz2lVLmwrkcZg
3cfd6brTclPTRxU8D+E8V95fqXLurfM1rZJ9tt380MezX/toYLwhdA1r26k7UR4P
iilscs/uld+uGYCSMjDMWHfWSR6Fz3ckGCdzr1yvMeKi2pbaakd4kF7ICwtDnPOs
BLJatB+aAulZ0bKJWYYdhwgCzqUMrgf0w5ojBJ51TnTMXXburtmSqjcppcDuU3Fy
sVJ4QncxUGBtC0e61hM4zKs88pu4yO72iBkiuTTRfZNoMhM0FZ++Thiw+PB8bULG
nSlQlN5CMllRHs1Yy3JzzTaLkKsFJ5B1/Frrg8yv8AI6+HMwxdN2H/+RA21Grvqi
EZtS9rRPgg8rMB/5cB53/1HPr7tQXtzO2fulFRoJwFEbAV0W0n2fzoFvNu7rpJYJ
hVkO+9Zr8xxCX80gpjplvyV3sK0Zw0G+jOELhxqXmPDyIqgLarhZ7a/tx6asYO+M
em97A1l/
=31o2
-----END PGP SIGNATURE-----
```


</details>
